language/move-lang/src/parser/ast.rs

```diff
@@ -0,0 +1,488 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::shared::{sp, Address, Identifier, Loc, Name, Spanned, TName};
+use std::fmt;
+
+macro_rules! new_name {
+    ($n:ident) => {
+        #[derive(Debug, Hash, Eq, PartialEq, Ord, PartialOrd, Clone)]
+        pub struct $n(pub Name);
+
+        impl TName for $n {
+            type Key = String;
+            type Loc = Loc;
+
+            fn drop_loc(self) -> (Loc, String) {
+                (self.0.loc, self.0.value)
+            }
+
+            fn clone_drop_loc(&self) -> (Loc, String) {
+                (self.0.loc, self.0.value.clone())
+            }
+
+            fn add_loc(loc: Loc, key: String) -> Self {
+                $n(sp(loc, key))
+            }
+        }
+
+        impl Identifier for $n {
+            fn value(&self) -> &str {
+                &self.0.value
+            }
+            fn loc(&self) -> Loc {
+                self.0.loc
+            }
+        }
+
+        impl fmt::Display for $n {
+            fn fmt(&self, f: &mut fmt::Formatter) -> std::fmt::Result {
+                write!(f, "{}", &self.0)
+            }
+        }
+    };
+}
+
+//**************************************************************************************************
+// Program
+//**************************************************************************************************
+
+#[derive(Debug)]
+pub struct Program {
+    pub source_definitions: Vec<FileDefinition>,
+    pub lib_definitions: Vec<FileDefinition>,
+}
+
+#[derive(Debug)]
+#[allow(clippy::large_enum_variant)]
+pub enum FileDefinition {
+    Modules(AddressDirective, Vec<ModuleDefinition>),
+    Main(Main),
+}
+
+#[derive(Debug)]
+pub enum AddressDirective {
+    Sender,
+    Specified(Loc, Address),
+}
+
+#[derive(Debug)]
+pub struct Main {
+    pub uses: Vec<(ModuleIdent, Option<ModuleName>)>,
+    pub function: Function,
+}
+
+//**************************************************************************************************
+// Modules
+//**************************************************************************************************
+
+new_name!(ModuleName);
+
+#[derive(Debug, Hash, Eq, PartialEq, Ord, PartialOrd, Clone)]
+pub struct ModuleIdent_ {
```

Comment:
this keeps being very odd to me. I would have done it the other way around which is carrying and optional location info for the ast node always. This side struct was very confusing (and ugly in my opinion) in the original ast but maybe you fixed the places where this appears. But effectively the idea the you have different types you are moving around and the fact that if you take the inner type you are hosed (you lost the correlation to source code) seemed always odd to me.
Maybe the pub on the structs with `_` fooled me or I don't know what....

Reply:
I guess we talked about this and you explained to me what I was missing. It sounds good! thanks